datahub-actions/src/datahub_actions/plugin/source/acryl/datahub_cloud_event_source.py

```diff
@@ -0,0 +1,202 @@
+
+ENTITY_CHANGE_EVENT_NAME = "entityChangeEvent"
+
+
+# Converts a DataHub Events Message to an EntityChangeEvent.
+def build_entity_change_event(payload: GenericPayloadClass) -> EntityChangeEvent:
+    return EntityChangeEvent.from_json(payload.get("value"))
+
+
+class DataHubEventsSourceConfig(ConfigModel):
+    topic: str = "PlatformEvent_v1"
+    consumer_id: Optional[str]  # Used to store offset for the consumer.
+    lookback_days: Optional[int] = None
+    reset_offsets: Optional[bool] = False
+
+    # Time and Exit Conditions.
+    kill_after_idle_timeout: bool = False
+    idle_timeout_duration_seconds: int = 30
+    event_processing_time_max_duration_seconds: int = 30
+
+
+# This is the custom DataHub-based Event Source.
+@dataclass
+class DataHubEventSource(EventSource):
+    running = False
+    source_config: DataHubEventsSourceConfig
+    ctx: PipelineContext
+
+    @staticmethod
+    def _get_pipeline_urn(pipeline_name: str) -> str:
+        if pipeline_name.startswith("urn:li:dataHubAction:"):
+            return pipeline_name
+        else:
+            return f"urn:li:dataHubAction:{pipeline_name}"
+
+    def __init__(self, config: DataHubEventsSourceConfig, ctx: PipelineContext):
+        self.ctx = ctx
+        self.source_config = config
+        self.consumer_id = DataHubEventSource._get_pipeline_urn(self.ctx.pipeline_name)
+
+        # Ensure a Graph Instance was provided.
+        assert self.ctx.graph is not None
+
+        self.datahub_events_consumer: DataHubEventsConsumer = DataHubEventsConsumer(
+            # TODO: This PipelineContext provides an Acryl Graph Instance
+            graph=self.ctx.graph.graph,
+            consumer_id=self.consumer_id,
+            lookback_days=self.source_config.lookback_days,
+            reset_offsets=self.source_config.reset_offsets,
+        )
+        self.ack_manager = AckManager()
+        self.safe_to_ack_offset: Optional[str] = None
+
+    @classmethod
+    def create(cls, config_dict: dict, ctx: PipelineContext) -> "EventSource":
+        config = DataHubEventsSourceConfig.parse_obj(config_dict)
+        return cls(config, ctx)
+
+    def events(self) -> Iterable[EventEnvelope]:
+        logger.info("Starting DataHub Cloud events source...")
+        logger.info(f"Subscribing to the following topic: {self.source_config.topic}")
+        self.running = True
+        yield from self._poll_and_process_events()
+
+    def _poll_and_process_events(self) -> Iterable[EventEnvelope]:
+        """Poll and process events in the main loop."""
+        last_idle_response_timestamp = 0
+        while self.running:
+            try:
+                sleeps_to_go = (
+                    self.source_config.event_processing_time_max_duration_seconds
+                )
+
+                while self.ack_manager.outstanding_acks():
+                    time.sleep(1)
+                    sleeps_to_go -= 1
+                    logger.info(f"Sleeps to go: {sleeps_to_go}")
```

Comment:
this seems like a logger.debug line at most? Wouldn't this crowd the event stream?

Reply:
yes! good catch